#160609514 Reset password by sending an email link

RonKbS opened the pull request:
#### What does this PR do?
Allow users to reset the password of their accounts.

#### Description of Task to be completed?
Currently, a user is unable to resent their password using the frontend of the application.

This PR allows a user to reset the password for their account by inputting the email. They then get a link via that email which they will be able to use to update the password.

#### How should this be manually tested?

Run the application using `npm start`
Click on `Forgot Password?` on the login page
Submit the `username` and `email` of a registered user.
A notification will pop up for checking the email address.
Using the link in the address, set a new password.
Use the new password details to login.

#### What are the relevant pivotal tracker stories?

[#160609514](https://www.pivotaltracker.com/story/show/160609514)



#### Any background context you want to add?
N/A


## Checklist:

- [x] My IDE is configured to follow the [**ESlint Conventions**](https://github.com/eslint/eslint)

- [x] I have **added tests** to cover my changes.

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the develop branch.

samuelbwambale reviewed (commented):
@RonKbS I was able to receive the resetPassword link however when I click on it, I get a response of page not found as seen below. Kindly look into this.

<img width="1439" alt="screen shot 2018-11-14 at 09 36 34" src="https://user-images.githubusercontent.com/31615608/48464646-4379f200-e7f1-11e8-8fab-11129c3696b4.png">

RonKbS commented:
@samuelbwambale you are interacting with the hosted app which does not have those changes so you'll have to change the URL in the axios instance to the local one on your pc.

samuelbwambale commented on src/actions/userActions.js:
```diff
@@ -115,3 +120,35 @@ export const getProfile = () => dispatch => {
+  .put('/api/users/password_update/', passwordDetails)
+  .then(response => {
+    dispatch({ type: RESET_PASSWORD_SUCCESS, payload: true });
+    toast.dismiss();
```
@RonKbS There is a `toast.dismiss` twice in this function. Consider refactoring it to the start of the function to cater for either successful response or an error.

samuelbwambale replied:
@RonKbS I would like to know if this was this fixed. Thanks.

RonKbS replied:
I had worked on the action before that and left this one.
Thanks for noticing @samuelbwambale.

RonKbS committed:
feature(authentication): reset password

- send a clickable link to email entered into a form
- enter password details to change password

MuhanguziDavid reviewed (commented):
@RonKbS , I cant seem to get the the reset password page.
<img width="1350" alt="screen shot 2018-11-14 at 15 02 00" src="https://user-images.githubusercontent.com/38077368/48481416-7b4b5e80-e81e-11e8-95c0-d37c18728b80.png">

kakaemma commented on src/actions/userActions.js:
```diff
@@ -130,3 +135,39 @@ export const getProfile = () => dispatch => {
+          { autoClose: 3500, hideProgressBar: true },
+        );
+      }
+      return toast.error('Error connection', { autoClose: 3500, hideProgressBar: true });
```
Thank you for handling errors. However, I think you should look into changing` Error connection` to either `Connection error` or some other message.

kakaemma commented on src/actions/userActions.js:
```diff
@@ -130,3 +135,39 @@ export const getProfile = () => dispatch => {
+  axiosInstance
+    .put('/api/users/password_update/', passwordDetails)
+    .then(response => {
+      dispatch({ type: RESET_PASSWORD_SUCCESS, payload: true });
```
I like the way you're making the dispatch. Wouldn't it be better if you had action creators and imported them here instead of hardcoding the actions?

RonKbS replied:
Since these are trivial action creators with boolean payloads, I think this should be ok.

dondrzzy replied:
I think @kakaemma makes a valid point. USing `action creators` makes the code more organized and also help prevents future devs from creating the same actions in other parts of the code if they are clearly defined in the action creators. @RonKbS

kakaemma commented on src/components/resetPassword/ResetPasswordPage.js:
```diff
@@ -0,0 +1,81 @@
+ResetPasswordPage.defaultProps = {
+  updatePasswordSuccess: false,
+};
+
```
consider removing line `67` and `72` that are being redundant.

kakaemma commented on src/components/resetPassword/ResetPasswordPage.js:
```diff
@@ -0,0 +1,81 @@
+    }
+    return toast.error('Make sure the password entered is the same', { autoClose: 3500, hideProgressBar: true });
+  };
+
```
This line is not used. kindly remove it.

kakaemma commented on src/components/resetPassword/ResetPasswordForm.js:
```diff
@@ -0,0 +1,54 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+// import { NavLink } from 'react-router-dom';
```
Kindly remove this line since the import since it is not useful here.

kakaemma commented on src/components/resetPassword/ResetPasswordForm.js:
```diff
@@ -0,0 +1,54 @@
+  password1: PropTypes.string.isRequired,
+  password2: PropTypes.string.isRequired,
+};
+
```
Line `52` is another redundant line in the file. Consider removing it

kakaemma commented on src/components/resetPassword/ResetPasswordForm.js:
```diff
@@ -0,0 +1,54 @@
+              <input name="recover-submit" className="btn btn-lg btn-primary btn-block" value="Reset" type="submit" onClick={onSave} />
+            </div>
+          </div>
+
```
I dont think we need an extra line here. so this also has to be removed.

kakaemma commented on src/tests/components/ForgotPassword.test.js:
```diff
@@ -0,0 +1,59 @@
+    wrapper.unmount();
+  });
+
+  it('should redirect on enteriing email details', () => {
```
Kindly consider changing the spelling for `enteriing`

kakaemma commented on src/tests/components/ResetPassword.test.js:
```diff
@@ -0,0 +1,88 @@
+    expect(spy).toBeCalled();
+  });
+
+  it('should not accept if passwords are not the same', () => {
```
@RonKbS, I do not understand this test. What shouldn't it accept?

RonKbS replied:
The test checks if the component rejects the form submission if the second input is invalid.

dondrzzy replied:
@RonKbS, your response to @kakaemma here seems more descriptive. Please consider using that instead. `it should not submit form data if the passwords are not the same`.

kakaemma commented on src/components/resetPassword/ResetPasswordPage.js:
```diff
@@ -0,0 +1,81 @@
+
+  componentDidMount() {
+    const { match: { params: { token } } } = this.props;
+    localStorage.setItem('auth_token', token);
```
Since successful login has`auth_token`set in local storage, setting it here would mean that a user can access all the protected routes before the password is actually reset. Kindly consider changing `auth_token` to another variable name.

kakaemma commented on src/components/resetPassword/ResetPasswordPage.js:
```diff
@@ -0,0 +1,81 @@
+
+  componentDidMount() {
+    const { match: { params: { token } } } = this.props;
+    localStorage.setItem('auth_token', token);
```
Since successful login has`auth_token`set in local storage, setting it here would mean that a user can access all the protected routes before the password is actually reset. Kindly consider changing `auth_token` to another variable name.

RonKbS replied:
This token is different from that of login so the user won't actually be able to access the other routes.
I will change the name though.

kakaemma replied:
@RonKbS, I tried it out and was able to access the dashboard.

RonKbS replied:
Update:
The dashboard is not protected.
Also, since I'm using the same Axios instance, I rely on the same key otherwise, as @MuhanguziDavid has experienced, the request is sent without a token if I change the key.
The token is removed after updating the password.

dondrzzy replied:
@RonKbS, please consider using a different local storage key. It is not an `auth_token`. Its a `reset_token` or something more suitable. What you save it as in the front-end does not matter as long you name it to the right key before sending it to the backend.

samuelbwambale reviewed (approved):
LGTM

dondrzzy commented on src/tests/components/ResetPassword.test.js:
```diff
@@ -0,0 +1,88 @@
+import React from 'react';
+import { mount } from 'enzyme';
+import { toast } from 'react-toastify';
+import { ResetPasswordPage } from '../../components/resetPassword/ResetPasswordPage';
```
Use ResetPassword instead of ResetPasswordPage.

dondrzzy reviewed (commented):
Well done, please refactor.

mendozabree reviewed (approved):
@RonKbS I like the new look of the email when the link is sent!
LGTM

MuhanguziDavid reviewed (commented):
Great work @RonKbS , however, I notice that I have to refresh the page on which I enter a new password for it to work (it seems to work on the second go). Could it be something to do with the state variables.

RonKbS commented:
@MuhanguziDavid it's to do with the axios instance we are all using. I've used axios as it is and set the authorization header in it to the token for resetting the password.

MuhanguziDavid reviewed (approved):
LGTM

dondrzzy commented on src/actions/userActions.js:
```diff
@@ -130,3 +135,41 @@ export const getProfile = () => dispatch => {
+    .catch((error) => {
+      if (error.response) {
+        dispatch(ResetPasswordError(error.response.data.errors[0]));
+        return toast.error(
```
Why return here when the overall result is not returned?

RonKbS replied:
My mistake. I was getting an eslint error of `a return is required` in some other sections. But yes this is not required so I will remove the return.

dondrzzy reviewed (commented):
Please refactor.

dondrzzy commented on src/tests/reducers/passwordResetReducers.test.js:
```diff
@@ -0,0 +1,111 @@
+    };
+    const currentState = userReducer(initialState, action);
+    expect(currentState).toEqual({
+      loading: false,
```
Everyone seems to be doing this which is making this file ambiguous. How about this:

expect(...).toEqual({
    ...initialState,
    resetPasswordloading: action.payload,
});
@RonKbS . If it works, advise the next victims.

dondrzzy commented on src/actions/userActions.js:
```diff
@@ -130,3 +135,41 @@ export const getProfile = () => dispatch => {
+    })
+    .catch((error) => {
+      if (error.response) {
+        dispatch(ResetPasswordError(error.response.data.errors[0]));
```
This will toast the error message and go ahead to toast the connection error all the time. Put the return statements back and return the axiosInstance.

RonKbS replied:
Oh yes,
I had forgotten that it will do that. 
Sorry for overlooking that.
I'm pushing the recommended changes now.

dondrzzy reviewed (commented):
Please refactor accordingly.
